core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ReplaceRedundantJoinWithSource.java

```diff
@@ -0,0 +1,111 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.iterative.rule;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.matching.Captures;
+import io.trino.matching.Pattern;
+import io.trino.sql.planner.iterative.Rule;
+import io.trino.sql.planner.plan.FilterNode;
+import io.trino.sql.planner.plan.JoinNode;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.sql.planner.iterative.rule.Util.restrictOutputs;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isAtMost;
+import static io.trino.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static io.trino.sql.planner.plan.Patterns.join;
+
+/**
+ * This rule transforms plans with joins, where:
+ * - one of the sources is scalar and produces no output symbols.
+ * In case of LEFT or RIGHT join, it has to be the inner source.
+ * In case of FULL or INNER join, it can be either source.
+ * - the other join source is at least scalar (not known to be empty).
+ * <p>
+ * The join is replaced with the other source and an optional
+ * pruning projection.
+ */
+public class ReplaceRedundantJoinWithSource
+        implements Rule<JoinNode>
+{
+    private static final Pattern<JoinNode> PATTERN = join();
+
+    @Override
+    public Pattern<JoinNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public Result apply(JoinNode node, Captures captures, Context context)
+    {
+        boolean leftSourceScalarWithNoOutputs = node.getLeft().getOutputSymbols().isEmpty() && isScalar(node.getLeft(), context.getLookup());
+        boolean rightSourceScalarWithNoOutputs = node.getRight().getOutputSymbols().isEmpty() && isScalar(node.getRight(), context.getLookup());
+
+        boolean leftSourceNotKnownEmpty = !isAtMost(node.getLeft(), context.getLookup(), 0);
+        boolean rightSourceNotKnownEmpty = !isAtMost(node.getRight(), context.getLookup(), 0);
+
+        switch (node.getType()) {
+            case INNER:
+                if (node.isCrossJoin()) {
+                    if (leftSourceScalarWithNoOutputs && rightSourceNotKnownEmpty) {
+                        return Result.ofPlanNode(restrictOutputs(context.getIdAllocator(), node.getRight(), ImmutableSet.copyOf(node.getRightOutputSymbols()))
+                                .orElse(node.getRight()));
+                    }
+                    if (leftSourceNotKnownEmpty && rightSourceScalarWithNoOutputs) {
+                        return Result.ofPlanNode(restrictOutputs(context.getIdAllocator(), node.getLeft(), ImmutableSet.copyOf(node.getLeftOutputSymbols()))
+                                .orElse(node.getLeft()));
+                    }
+                }
+                else {
+                    if (leftSourceScalarWithNoOutputs && rightSourceNotKnownEmpty) {
+                        checkState(node.getFilter().isPresent());
```

Comment:
This is unnecessary. The call to node.getFilter().get() will fail in a similar way if the filter is missing.

Reply:
IntelliJ makes me do this... Otherwise i get a warning.